### ISSUE TITLE
Limit amount of tablets globally

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/scylladb#21458]
Having too many tablets causes all sorts of issues:

- Excessive metadata footprint / OOM
  - due to `tablet_map` size in memory
  - compaction group overhead, sstable metadata overhead (more relevant)  Refs #20959
- making group0 unoperational:
  - https://github.com/scylladb/scylladb/issues/21456

We should put safety limits in the tablet allocator to avoid:
  1) excessive global tablet count (metadata size issues)
  2) excessive per-shard tablet count (compaction-group count per shard related issues)

To avoid (2), we should trigger tablet merge to keep the per-shard count in check on scale down.

**Comment by raphaelsc** [comment]
> Having too many tablets causes all sorts of issues:
> 
> * Excessive metadata footprint / OOM
>   
>   * due to `tablet_map` size in memory
>   * compaction group overhead, sstable metadata overhead (more relevant)  Refs [terminate called after throwing an instance of 'std::bad_alloc' on commilog replay with high number of initial tablets and high number of tables #20959](https://github.com/scylladb/scylladb/issues/20959)
> * making group0 unoperational:
>   
>   * [Large number of tablets can break group0 ability to send snapshots #21456](https://github.com/scylladb/scylladb/issues/21456)
> 
> We should put safety limits in the tablet allocator to avoid:
> 
> 1. excessive global tablet count (metadata size issues)
> 2. excessive per-shard tablet count (compaction-group count per shard related issues)
> 
> To avoid (2), we should trigger tablet merge to keep the per-shard count in check on scale down.

Good point. Theoretically, there could be a rare TTLd workload which is constantly creating new tables and retiring old ones. Old ones remain with a large tablet count (result of ingestion + splits) while eventually having no data at all (due to expiring data). So global tablet count will only grow over time until we hit these issues. Tablet merge will fix this though. For fixing problem 1, we may have to adjust the target tablet size to respect the global tablet count goal, it's a trade-off in the end.
